marconi-core/src/Marconi/Core/Indexer/FileIndexer.hs

```diff
@@ -352,8 +360,42 @@ writeStable
 writeStable p indexer = do
   let f = lastStableFilename indexer
       serialise = indexer ^. fileBuilder . serialisePoint
-  liftIO $ BS.writeFile f $ serialise p
+  writeIndexer indexer f $ serialise p
   pure indexer
 
-instance (Applicative m) => Closeable m (FileIndexer meta) where
-  close = const $ pure ()
+instance (MonadIO m) => Closeable m (FileIndexer meta) where
+  {- If we've got a write token it means a write is occurring asynchronously.
+     We need to wait for it to finish before closing. -}
+  close :: FileIndexer meta event -> m ()
+  close indexer = liftIO $
+    case indexer ^. fileIndexerWriteTokens of
+      Just sem -> Con.waitQSemN sem 1
+      Nothing -> pure ()
+
+-- * File writing
+
+-- TODO maybe this section goes into its own file with `BS.writeFile` abstracted so we can use it
+--      for DB inserts or whatever
+writeFileWith :: (MonadIO m) => (IO () -> IO ()) -> FilePath -> ByteString -> m ()
+writeFileWith executor filename content =
+  liftIO $ executor (BS.writeFile filename content)
+
+writeFileAsync :: (MonadIO m) => QSemN -> FilePath -> ByteString -> m ()
+writeFileAsync qsem = writeFileWith (flip withAsync $ const (liftIO $ waitQSemN qsem 1))
```

Comment:
It looks like we aren't releasing the lock here (via `signalQSem`/`signalQSemN`): you need to `waitQSem` before writing and to `signalQSem` once it's done.
So, probably acquiring the lock in the `withAsync` action and releasing it afterwards.
Also, in the 2nd parameter of `withAsync`, you probably want to `wait` for the result, otherwise, you'll just run the action without even waiting for the end of the writing.

Reply:
Thanks, I didn't understand exactly how it works yet, this makes it clear!